src/ast.rs

```diff
@@ -202,38 +204,165 @@ pub enum RawUnaryOperator {
+/// <https://google.github.io/xls/dslx_reference/#concat-expression>
+///
+/// They are ordered from highest precedence to lowest, and grouped when same precedence.
+#[derive(Debug, PartialEq, Clone, Copy)]
 pub enum RawBinaryOperator {
-    /// `||`, boolean or
-    BooleanOr,
-    /// `&&`, boolean and
-    BooleanAnd,
-
-    /// `|`, bit-wise or
-    BitwiseOr,
-    /// `&`, bit-wise and
-    BitwiseAnd,
-    /// `^`, bit-wise xor
-    BitwiseXor,
+    /// `*`, multiply
+    Multiply,
 
     /// `+`, add
     Add,
     /// `-`, subtract
     Subtract,
-    /// `*`, multiply
-    Multiply,
+    /// `++` bitwise concatenation
+    Concatenate,
 
     /// `>>`, shift right (both logical and arithmetic, depending on context)
     ShiftRight,
     /// `<<`, shift left
     ShiftLeft,
+
+    /// `&`, bit-wise and
+    BitwiseAnd,
+
+    /// `^`, bit-wise xor
+    BitwiseXor,
+
+    /// `|`, bit-wise or
+    BitwiseOr,
+
+    /// `==` equal
+    Equal,
+    /// `!=` not equal
+    NotEqual,
+    /// `>=` greater or equal
+    GreaterOrEqual,
+    /// `>`  greater
+    Greater,
+    /// `<=` less or equal
+    LessOrEqual,
+    /// `<`  less
+    Less,
+
+    /// `&&`, boolean and
+    BooleanAnd,
+
+    /// `||`, boolean or
+    BooleanOr,
 }
 
 pub type BinaryOperator = Spanned<RawBinaryOperator>;
 
+impl RawBinaryOperator {
+    /// Returns true for the 6 comparison operators (== != < > <= >=), otherwise false.
+    fn is_comparison(&self) -> bool {
+        match self {
+            RawBinaryOperator::Multiply => false,
+            RawBinaryOperator::Add => false,
+            RawBinaryOperator::Subtract => false,
+            RawBinaryOperator::Concatenate => false,
+            RawBinaryOperator::ShiftRight => false,
+            RawBinaryOperator::ShiftLeft => false,
+            RawBinaryOperator::BitwiseAnd => false,
+            RawBinaryOperator::BitwiseXor => false,
+            RawBinaryOperator::BitwiseOr => false,
+            RawBinaryOperator::Equal => true,
+            RawBinaryOperator::NotEqual => true,
+            RawBinaryOperator::GreaterOrEqual => true,
+            RawBinaryOperator::Greater => true,
+            RawBinaryOperator::LessOrEqual => true,
+            RawBinaryOperator::Less => true,
+            RawBinaryOperator::BooleanAnd => false,
+            RawBinaryOperator::BooleanOr => false,
+        }
+    }
+}
+
+impl PartialOrd for RawBinaryOperator {
+    /// Implements DSLX binary operator precedence. I.e. x Some(Greater) y means that x is
+    /// higher in the precedence table than y. x None y means they're the same precedence.
+    /// see <https://google.github.io/xls/dslx_reference/#operator-precedence>
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        if PartialEq::eq(self, other) {
+            Some(Ordering::Equal)
+        } else {
+            match (self, other) {
+                (Self::Multiply, _) => Some(Ordering::Greater),
```

Comment:
Divide & modulo & IDK what else coming later?

Reply:
Yes, those are coming later. I was following https://google.github.io/xls/dslx_reference/#binary-expressions , and divide and modulo were not mentioned... but I/we won't forget about them.